core/services/directrequestocr/orm.go

```diff
@@ -19,121 +19,90 @@ type ORM interface {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		val.RunID = runID
-		val.ErrorType = errorType
-		val.Error = computationError
-		val.State = RESULT_READY
-		val.Result = []byte{}
-		val.ResultReadyAt = readyAt
-		o.db[requestID] = val
-		return nil
+func (o *orm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
+	var oldState RequestState
+
+	stmt := `SELECT state FROM ocr2dr_requests WHERE request_id=$1;`
+	if err := o.q.WithOpts(qopts...).Get(&oldState, stmt); err != nil {
+		return 0, err
 	}
-	return fmt.Errorf("can't find entry with requestID: %v", requestID)
-}
 
-func (o *inmemoryorm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	prevState := IN_PROGRESS
-	if val, ok := o.db[requestID]; ok {
-		prevState = val.State
-		val.State = state
-		o.db[requestID] = val
-		return prevState, nil
+	if oldState == state {
+		return state, nil
 	}
-	return prevState, fmt.Errorf("can't find entry with requestID: %v", requestID)
+
+	stmt = `UPDATE ocr2dr_requests SET (state=$2) WHERE request_id=$1;`
+	return oldState, o.q.WithOpts(qopts...).ExecQ(stmt, requestID, state)
 }
 
-func (o *inmemoryorm) FindOldestEntriesByState(state RequestState, limit uint32, qopts ...pg.QOpt) ([]Request, error) {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	var result []Request
-	// NOTE: suboptimal if limit << full result
-	for _, val := range o.db {
-		if val.State == state {
-			result = append(result, val)
-		}
-	}
-	sort.Slice(result, func(i, j int) bool {
-		return result[i].ReceivedAt.Before(result[j].ReceivedAt)
-	})
-	if limit < uint32(len(result)) {
-		result = result[:limit]
-	}
-	return result, nil
+func (o *orm) SetTransmittedResult(requestID RequestID, transmittedResult []byte, transmittedError []byte, qopts ...pg.QOpt) error {
+	stmt := `
+		UPDATE ocr2dr_requests
+		SET (transmitted_result=$2, transmitted_error=$3)
+		WHERE request_id=$1;
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID, transmittedResult, transmittedError)
 }
 
-func (o *inmemoryorm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		return &val, nil
-	}
-	return nil, fmt.Errorf("can't find entry with dbid: %v", requestID)
+func (o *orm) FindOldestEntriesByState(state RequestState, limit uint32, qopts ...pg.QOpt) ([]Request, error) {
+	var requests []Request
+	stmt := `SELECT * FROM ocr2dr_requests WHERE state=$1 ORDER_BY received_at DESC LIMIT $2;`
+	err := o.q.WithOpts(qopts...).Get(&requests, stmt, state, limit)
+	return requests, err
 }
 
-// TODO actual DB: https://app.shortcut.com/chainlinklabs/story/54049/database-table-in-core-node
+func (o *orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
+	request := &Request{}
+	stmt := `SELECT * FROM ocr2dr_requests WHERE request_id=$1;`
```

Comment:
Does it make sense to always include contract ID in SELECTs? I guess we're depending on the practical impossibility of having duplicate request IDs... I wonder if checking contract ID could help us with potential rare bugs where reqIDs get messed up...